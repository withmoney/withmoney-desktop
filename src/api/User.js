import axios from 'axios';

export const login = data => axios.post('http://localhost:3000/api/v1/login', data);

export const signup = data => axios.post('http://localhost:3000/api/v1/signup', data);
