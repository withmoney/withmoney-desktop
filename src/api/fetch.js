import axios from 'axios';
import { catchInvalidToken } from 'app/utils/parse';

const newAxios = axios.create({
  baseURL: `${process.env.MYMONEY_API}`,
  headers: {
    Authorization: {
      toString() {
        return `Bearer ${localStorage.getItem('token')}`;
      },
    },
  },
});

const fetch = method => (...args) => catchInvalidToken(newAxios[method](...args));

export default {
  get: fetch('get'),
  post: fetch('post'),
  put: fetch('put'),
  delete: fetch('delete'),
};
