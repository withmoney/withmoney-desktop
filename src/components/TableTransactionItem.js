import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import classnames from 'classnames';
import * as TransactionsActions from 'store/transactions';
import InputInline from 'components/InputInline';
import If from 'components/render-utils/If';
import ButtonRounded from 'components/ButtonRounded';

class TransactionsItem extends React.Component {
  constructor(props) {
    super(props);

    this.toggleEditing = this.toggleEditing.bind(this);
    this.onDoubleClick = this.onDoubleClick.bind(this);
    this.handleInput = this.handleInput.bind(this);
    this.save = this.save.bind(this);

    this.fields = [
      { name: 'transactionDate' },
      { name: 'name' },
      { name: 'CategoryId' },
      { name: 'value' },
    ];

    this.state = {
      isEditing: false,
      formData: props.transaction,
    };
  }

  onDoubleClick() {
    const { isEditing } = this.state;

    if (!isEditing) {
      this.toggleEditing();
    }
  }

  handleInput({ target: { name, value } }) {
    const { formData } = this.state;
    this.setState({
      formData: {
        ...formData,
        [name]: value,
      },
    });
  }

  async save() {
    const { formData } = this.state;
    const {
      actions: { transaction },
    } = this.props;

    try {
      await transaction.put(formData);

      this.toggleEditing();
    } catch (error) {
      console.error(error);
    }
  }

  toggleEditing() {
    const { isEditing } = this.state;
    const { transaction: formData } = this.props;

    if (!isEditing) {
      this.setState({ formData });
    }

    this.setState({
      isEditing: !isEditing,
    });
  }

  renderAction(classCol) {
    const { isEditing } = this.state;
    const {
      transaction: { isLoading },
    } = this.props;

    return (
      <Fragment>
        <If condition={isEditing}>
          <div className={classCol}>
            <ButtonRounded type="button" onClick={this.save} medium isLoading={isLoading}>
              Save
            </ButtonRounded>
          </div>
        </If>
        <If condition={!isEditing}>
          <div className={classCol}>
            <input type="checkbox" />
          </div>
        </If>
      </Fragment>
    );
  }

  render() {
    const { isEditing, formData } = this.state;
    const {
      transaction: { isLoading },
    } = this.props;
    const classCol = classnames('table-transactions__col', {
      'table-transactions__col--is-editing': isEditing,
    });

    return (
      <div
        className={classnames('table-transactions__row', {
          'table-transactions__row--is-editing': isEditing,
        })}
        onDoubleClick={this.onDoubleClick}
      >
        {this.fields.map(field => (
          <div key={field.name} className={classCol}>
            <InputInline
              isEditing={isEditing}
              name={field.name}
              className="table-transactions__input"
              defaultValue={formData[field.name]}
              disabled={isLoading}
              onChange={this.handleInput}
            />
          </div>
        ))}
        {this.renderAction(classCol)}
      </div>
    );
  }
}

TransactionsItem.propTypes = {
  actions: PropTypes.shape({
    transaction: PropTypes.shape({
      put: PropTypes.func.isRequired,
    }).isRequired,
  }).isRequired,
  transaction: PropTypes.shape({
    id: PropTypes.number,
    transactionDate: PropTypes.string,
    name: PropTypes.string,
    CategoryId: PropTypes.number,
    value: PropTypes.string,
    isLoading: PropTypes.bool,
  }),
};

TransactionsItem.defaultProps = {
  transaction: {
    isLoading: false,
  },
};

const mapDispachToProps = dispatch => ({
  actions: {
    transaction: bindActionCreators(TransactionsActions, dispatch),
  },
});

export default connect(
  null,
  mapDispachToProps,
)(TransactionsItem);
