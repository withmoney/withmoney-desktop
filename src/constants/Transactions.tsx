import { TransactionType } from '../models';

export const addOperationText = {
  [TransactionType.Deposit]: 'Add Entrance',
  [TransactionType.FixedExpense]: 'Add Recurrent',
  [TransactionType.CreditCard]: 'Add Credit',
  [TransactionType.VariableExpense]: 'Add Unforeseen',
};

export const operationType = [
  [TransactionType.Deposit],
  [TransactionType.FixedExpense],
  [TransactionType.CreditCard],
  [TransactionType.VariableExpense],
];
