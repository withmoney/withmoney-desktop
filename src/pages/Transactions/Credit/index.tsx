import React from 'react';
import Container from '../Components/Container';
import { Buttons } from '../Components/Buttons';

const Credit = () => {
  return (
    <Container>
      <Buttons activeButton="Credit" />
    </Container>
  );
};

export default Credit;
