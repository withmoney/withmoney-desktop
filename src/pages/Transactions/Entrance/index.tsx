import React from 'react';
import styled from 'styled-components';
import { Buttons } from '../Components/Buttons';
import { Table, TableConfig } from '../Components/Table';
import Container from '../Components/Container';
import EntranceButton from './EntranceButton';
import EntranceTitle from './EntranceTitle';
import EntranceData from './EntranceData';
import EntranceLoad from './EntranceLoad';
import EntranceInfo from './EntranceInfo';

const Entrance = () => {
  return (
    <Container>
      <Buttons activeButton="Entrance" />
      <OperationContainer>
        <Table>
          <TableConfig />
          <EntranceTitle />
          <EntranceData />
          <EntranceLoad />
        </Table>
        <EntranceButton />
      </OperationContainer>
      <EntranceInfo />
    </Container>
  );
};

const OperationContainer = styled.div`
  display: flex;
  margin-bottom: 15px;
  flex-direction: column;
  padding: 30px;
  background-color: #ffff;
`;

export default Entrance;
