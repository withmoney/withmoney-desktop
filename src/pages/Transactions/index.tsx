import React from 'react';
import styled from 'styled-components';
import { Plus } from '@styled-icons/evaicons-solid';
import { Tabs } from './Components/Tabs';
import Table from './Components/Table';
import Button from '../../components/Button';
import Container from './Components/style/Container.style';
import EntranceTitle from './Operation/OperationTitle';
import EntranceData from './Operation/OperationData';
import EntranceLoad from './Operation/TransactionPlaceholder';
import FooterContainer from './Operation/FooterContainer';

const Operation = () => {
  return (
    <Container>
      <Tabs />
      <OperationContainer>
        <Table.Table>
          <Table.Config />
          <EntranceTitle />
          <EntranceData />
          <EntranceLoad />
        </Table.Table>
        <ButtonContent>
          <Button rounded variation="light">
            <Plus />
            <span>Add Entrance</span>
          </Button>
        </ButtonContent>
      </OperationContainer>
      <FooterContainer />
    </Container>
  );
};

const OperationContainer = styled.div`
  display: flex;
  margin-bottom: 15px;
  flex-direction: column;
  padding: 30px;
  background-color: #ffff;
`;

const ButtonContent = styled.div`
  display: flex;
  justify-content: center;
`;

export default Operation;
