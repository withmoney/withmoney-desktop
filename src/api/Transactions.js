import fetch from 'api/fetch';
import { mountQuery } from 'app/utils/parse';

export const list = (query = {}) => fetch.get(mountQuery('transactions', query));

export const get = id => fetch.get(`transactions/${id}`);

export const put = (id, data) => fetch.put(`transactions/${id}`, data);
