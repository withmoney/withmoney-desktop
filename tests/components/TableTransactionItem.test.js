import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import TableTransactionItem from '../../src/components/TableTransactionItem';
import * as Transactions from '../../src/api/Transactions';

const mockStore = configureStore([thunk]);

jest.mock('../../src/api/Transactions', () => ({
  put: jest.fn().mockResolvedValue({}),
}));

describe('TableTransactionItem', () => {
  let Component;

  beforeAll(() => {
    Component = global.withRedux(mockStore({}), TableTransactionItem);
  });

  it('should render a static row with transaction', () => {
    const wrapper = renderer.create(Component({
      transaction: {
        id: 55,
        transactionDate: '2019-05-31',
        name: 'Name',
        CategoryId: 4,
        value: '40.5',
        isLoading: false,
      },
    }));

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('Should edit transaction and quite when click on save button', (done) => {
    const wrapper = mount(Component({
      transaction: {
        id: 55,
        transactionDate: '2019-05-31',
        name: 'Name',
        CategoryId: 4,
        value: '40.5',
      },
    }));

    expect(wrapper.find('TransactionsItem').state().isEditing).toBe(false);

    wrapper.find('.table-transactions__row').simulate('doubleclick');

    expect(wrapper.find('TransactionsItem').state().isEditing).toBe(true);

    wrapper.find('input[name="name"]').simulate('change', { target: { name: 'name', value: 'Name New' } });

    wrapper.find('button').simulate('click');

    expect(Transactions.put).toBeCalledWith(55, {
      id: 55,
      transactionDate: '2019-05-31',
      name: 'Name New',
      CategoryId: 4,
      value: '40.5',
    });

    setImmediate(() => {
      expect(wrapper.find('TransactionsItem').state().isEditing).toBe(false);
      done();
    });
  });
});
