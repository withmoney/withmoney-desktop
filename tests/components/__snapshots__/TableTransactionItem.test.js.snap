// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableTransactionItem should render a static row with transaction 1`] = `
<div
  className="table-transactions__row"
  onDoubleClick={[Function]}
>
  <div
    className="table-transactions__col"
  >
    2019-05-31
  </div>
  <div
    className="table-transactions__col"
  >
    Name
  </div>
  <div
    className="table-transactions__col"
  >
    4
  </div>
  <div
    className="table-transactions__col"
  >
    40.5
  </div>
  <div
    className="table-transactions__col"
  >
    <input
      type="checkbox"
    />
  </div>
</div>
`;

exports[`TableTransactionItem should render table in mode edition 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TransactionsItem)
    transaction={
      Object {
        "CategoryId": 1,
        "id": 99,
        "name": "Name",
        "transactionDate": "2018-08-01 20:80",
        "value": "0.00",
      }
    }
  >
    <TransactionsItem
      actions={
        Object {
          "transaction": Object {
            "default": [Function],
            "list": [Function],
            "put": [Function],
          },
        }
      }
      transaction={
        Object {
          "CategoryId": 1,
          "id": 99,
          "name": "Name",
          "transactionDate": "2018-08-01 20:80",
          "value": "0.00",
        }
      }
    >
      <div
        className="table-transactions__row table-transactions__row--is-editing"
        onDoubleClick={[Function]}
      >
        <div
          className="table-transactions__col table-transactions__col--is-editing"
          key="transactionDate"
        >
          <InputInline
            className="table-transactions__input"
            defaultValue="2018-08-01 20:80"
            isEditing={true}
            isHTML={false}
            name="transactionDate"
            onChange={[Function]}
            type="text"
          >
            <input
              className="table-transactions__input"
              defaultValue="2018-08-01 20:80"
              name="transactionDate"
              onChange={[Function]}
              type="text"
            />
          </InputInline>
        </div>
        <div
          className="table-transactions__col table-transactions__col--is-editing"
          key="name"
        >
          <InputInline
            className="table-transactions__input"
            defaultValue="Name"
            isEditing={true}
            isHTML={false}
            name="name"
            onChange={[Function]}
            type="text"
          >
            <input
              className="table-transactions__input"
              defaultValue="Name"
              name="name"
              onChange={[Function]}
              type="text"
            />
          </InputInline>
        </div>
        <div
          className="table-transactions__col table-transactions__col--is-editing"
          key="CategoryId"
        >
          <InputInline
            className="table-transactions__input"
            defaultValue={1}
            isEditing={true}
            isHTML={false}
            name="CategoryId"
            onChange={[Function]}
            type="text"
          >
            <input
              className="table-transactions__input"
              defaultValue={1}
              name="CategoryId"
              onChange={[Function]}
              type="text"
            />
          </InputInline>
        </div>
        <div
          className="table-transactions__col table-transactions__col--is-editing"
          key="value"
        >
          <InputInline
            className="table-transactions__input"
            defaultValue="0.00"
            isEditing={true}
            isHTML={false}
            name="value"
            onChange={[Function]}
            type="text"
          >
            <input
              className="table-transactions__input"
              defaultValue="0.00"
              name="value"
              onChange={[Function]}
              type="text"
            />
          </InputInline>
        </div>
        <If
          condition={true}
        >
          <div
            className="table-transactions__col table-transactions__col--is-editing"
          >
            <ButtonRounded
              isLoading={false}
              medium={true}
              onClick={[Function]}
              type="button"
            >
              <button
                className="button-rounded button-rounded--medium"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <If
                  condition={false}
                />
                <If
                  condition={true}
                >
                  Save
                </If>
              </button>
            </ButtonRounded>
          </div>
        </If>
        <If
          condition={false}
        />
      </div>
    </TransactionsItem>
  </Connect(TransactionsItem)>
</Provider>
`;
